crates/symbolicator-service/src/download/partial.rs

```diff
@@ -0,0 +1,311 @@
+            .headers()
+            .get(reqwest::header::CONTENT_RANGE)
+            .and_then(|hv| hv.to_str().ok())
+            .map(|s| s.parse())
+    }
+}
+
+/// An error which can be returned when parsing a [`BytesContentRange`].
+#[derive(thiserror::Error, Debug)]
+pub enum InvalidBytesRange {
+    /// The header is malformed.
+    #[error("content range header is malformed")]
+    Malformed,
+    /// The specified unit is not `bytes`.
+    #[error("content range unit is not bytes")]
+    NotBytes,
+    /// The header indicates the requested range is not satisfiable.
+    #[error("the requested range is not satisfiable")]
+    RangeNotSatisfiable,
+    /// The header indicates the total length of the resource is unknown.
+    #[error("the total length of the resource is not known")]
+    UnknownLength,
+}
+
+impl FromStr for BytesContentRange {
+    type Err = InvalidBytesRange;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let Some(("bytes", s)) = s.trim().split_once(' ') else {
+            return Err(InvalidBytesRange::NotBytes);
+        };
+
+        let (range, total_size) = s
+            .trim()
+            .split_once('/')
+            .ok_or(InvalidBytesRange::Malformed)?;
+        if range.trim() == "*" {
+            return Err(InvalidBytesRange::RangeNotSatisfiable);
+        }
+
+        let (start, end) = range.split_once('-').ok_or(InvalidBytesRange::Malformed)?;
+        let start = start
+            .trim()
+            .parse()
+            .map_err(|_| InvalidBytesRange::Malformed)?;
+        let end = end
+            .trim()
+            .parse()
+            .map_err(|_| InvalidBytesRange::Malformed)?;
+
+        if end < start {
+            return Err(InvalidBytesRange::Malformed);
+        }
+
+        let total_size = match total_size.trim() {
+            "*" => Err(InvalidBytesRange::UnknownLength),
+            size => size.parse().map_err(|_| InvalidBytesRange::Malformed),
+        }?;
+
+        Ok(Self {
+            start,
+            end,
+            total_size,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! assert_split {
+        ($r:expr, @$expected:literal) => {{
+            let s = split($r).collect::<Vec<_>>();
+            insta::assert_debug_snapshot!(s, @$expected);
+
+            if let Some(x) = s.last() {
+                assert_eq!($r.total_size, x.end + 1);
+            }
+
+            s
+        }};
+    }
+
+    #[test]
+    fn test_split_nothing_to_do() {
+        let r = BytesContentRange {
+            start: 0,
+            end: 11,
+            total_size: 12,
+        };
+
+        assert_split!(r, @"[]");
+    }
+
+    #[test]
+    fn test_split_too_small() {
+        let r = BytesContentRange {
+            start: 0,
+            end: 9,
+            total_size: 10 + PARTIAL_SPLIT_SIZE,
+        };
+
+        let s = split(r).collect::<Vec<_>>();
+        insta::assert_debug_snapshot!(s, @r###"
```

Comment:
Why the three `#`s here? Shouldn't one suffice?

Reply:
`cargo insta review` just generates that